contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,238 @@
+contract ERC2771Forwarder is EIP712, Nonces {
+    using ECDSA for bytes32;
+
+    struct ForwardRequestData {
+        address from;
+        address to;
+        uint256 value;
+        uint256 gas;
+        uint48 deadline;
+        bytes data;
+        bytes signature;
+    }
+
+    bytes32 private constant _FORWARD_REQUEST_TYPEHASH =
+        keccak256(
+            "ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,uint48 deadline,bytes data)"
+        );
+
+    /**
+     * @dev Emitted when a `ForwardRequest` is executed.
+     */
+    event ExecutedForwardRequest(address indexed signer, uint256 nonce, bool success, bytes returndata);
+
+    /**
+     * @dev The request `from` doesn't match with the recovered `signer`.
+     */
+    error ERC2771ForwarderInvalidSigner(address signer, address from);
+
+    /**
+     * @dev The requested `value` doesn't match with the available `msgValue`, leaving ETH stuck in the contract.
+     */
+    error ERC2771ForwarderMismatchedValue(uint256 value, uint256 msgValue);
+
+    /**
+     * @dev The list of requests length doesn't match with the list of signatures length.
+     */
+    error ERC2771ForwarderInvalidBatchLength(uint256 requestsLength, uint256 signaturesLength);
+
+    /**
+     * @dev The request `deadline` has expired.
+     */
+    error ERC2771ForwarderExpiredRequest(uint48 deadline);
+
+    /**
+     * @dev See {EIP712-constructor}.
+     */
+    constructor(string memory name) EIP712(name, "1") {}
+
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     *
+     * NOTE: A request signed with an invalid nonce will return false here but won't revert by default
+     * to prevent revert when a batch includes a request already executed by another relay. This behavior can be opt-out.
+     * See {executeBatch}.
+     */
+    function verify(ForwardRequestData calldata request) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, ) = _validate(request, nonces(request.from));
+        return alive && signerMatch;
+    }
+
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is left stuck in the contract.
+     */
+    function execute(ForwardRequestData calldata request) public payable virtual returns (bool, bytes memory) {
+        if (msg.value != request.value) {
+            revert ERC2771ForwarderMismatchedValue(request.value, msg.value);
+        }
+
+        (bool success, bytes memory returndata) = _execute(request, _useNonce(request.from), true);
+
+        return (success, returndata);
+    }
+
+    /**
+     * @dev Batch version of {execute} that also includes an `atomic` parameter to indicate whether all requests are
+     * executed or none of them.
+     */
+    function executeBatch(ForwardRequestData[] calldata requests, bool atomic) public payable virtual {
```

Comment:
I think we need to clearly document what the atomic boolean does, and what it doesn't do. 

I see the point for a relayer to use "false", so that the the rest of the batch passes if one of the request is invalid/expired/...

If you use true, you don't have such a clear upside. Yes your tx will fail if one the the request is not alive/signed ... but its still possible for some attacker to partially execute the requests. Said otherwise, atomic is a "best effort" that this tx will try do stuff atomically, but not a security guarantee that an attacker can't do a partial execution.

Reply:
I agree, I increased the documentation around this and added notes across the contract to indicate how the flag affects other values (eg. those emitted in an event). I also added a note indicating that the atomicity is only guaranteed for the first-level subcalls.